src/Chapter3.hs

```diff
@@ -1006,6 +1170,143 @@ Implement data types and typeclasses, describing such a battle between two
 contestants, and write a function that decides the outcome of a fight!
 -}
 
+data Action
+  = Attack
+  | DrinkPotion
+  | CastSpell
+  | RunAway deriving (Show)
+
+data Monster = Monster
+  { monsterHealth  :: Int
+  , monsterAttack  :: Int
+  , monsterActions :: [Action]
+  } deriving (Show)
+
+data Knight = Knight
+  { knightHealth  :: Int
+  , knightAttack  :: Int
+  , knightDefense :: Int
+  , knightActions :: [Action]
+  , knightSpell   :: Int
+  , knightPotion  :: Int
+  } deriving (Show)
+
+class Fighter a where
+  takeDamage :: a -> Int -> a
+  heal       :: a -> Int -> a
+  applySpell :: a -> Int -> a
+  isDead     :: a -> Bool
+  attack     :: (Fighter b) => a -> b -> b
+  doAction   :: (Fighter b) => a -> b -> Action -> (a, b)
+  doActions  :: (Fighter b) => a -> b -> (a, b)
+
+instance Fighter Monster where
+  takeDamage :: Monster -> Int -> Monster
+  takeDamage monster amount = monster { monsterHealth = monsterHealth monster - amount }
+
+  -- Monsters can't do this
+  heal :: Monster -> Int -> Monster
+  heal monster _ = monster
+
+  -- Monsters can't do this
+  applySpell :: Monster -> Int -> Monster
+  applySpell monster _ = monster
+
+  isDead :: Monster -> Bool
+  isDead = (<= 0) . monsterHealth
+
+  attack :: (Fighter b) => Monster -> b -> b
+  attack monster opponent = takeDamage opponent $ monsterAttack monster
+
+  doAction :: (Fighter b) => Monster -> b -> Action -> (Monster, b)
+  doAction monster opponent RunAway =
+    ( monster
+    , opponent
+    )
+  doAction monster opponent Attack =
+    ( monster
+    , takeDamage opponent $ monsterAttack monster
+    )
+
+  doActions :: (Fighter a) => Monster -> a -> (Monster, a)
+  doActions monster opponent = head $ map (doAction monster opponent) $ monsterActions monster
+
+instance Fighter Knight where
+  takeDamage :: Knight -> Int -> Knight
+  takeDamage knight amount = knight
+    { knightHealth = knightHealth knight - amount + knightDefense knight
+    , knightDefense = maximum [knightDefense knight - amount, 0]
+    }
+
+  heal :: Knight -> Int -> Knight
+  heal knight amount = knight
+    { knightHealth = knightHealth knight + amount
+    }
+
+  applySpell :: Knight -> Int -> Knight
+  applySpell knight amount = knight
+    { knightDefense = knightDefense knight + amount
+    }
+
+  isDead :: Knight -> Bool
+  isDead = (<= 0) . knightHealth
+
+  attack :: (Fighter a) => Knight -> a -> a
+  attack knight opponent = takeDamage opponent $ knightAttack knight
+
+  doAction :: (Fighter a) => Knight -> a -> Action -> (Knight, a)
+  doAction knight opponent Attack =
+    ( knight
+    , takeDamage opponent $ knightAttack knight
+    )
+  doAction knight opponent DrinkPotion =
+    ( heal knight $ knightPotion knight
+    , opponent
+    )
+  doAction knight opponent CastSpell =
+    ( applySpell knight $ knightSpell knight
+    , takeDamage opponent $ knightAttack knight
+    )
```

Comment:
To avoid repeating top-level function definition in such cases, it's usually recommended to use the `case-of` expression:

```haskell
doAction knight opponent action = case action of
    Attack -> ...
    DrinkPotion -> ...
    ...
```

Reply:
Thanks! I made an update, not sure if I handled `Either` properly. Please let me know :)